components/tikv_util/src/mpsc/future.rs

```diff
@@ -0,0 +1,471 @@
+            == queue
+                .liveness
+                .fetch_sub(RECEIVER_COUNT_BASE, Ordering::Release)
+        {
+            atomic::fence(Ordering::Acquire);
+            drop(unsafe { Box::from_raw(self.queue) });
+        }
+    }
+}
+
+unsafe impl<T: Send> Send for Receiver<T> {}
+
+pub fn unbounded<T>(policy: WakePolicy) -> (Sender<T>, Receiver<T>) {
+    let queue = Box::into_raw(Box::new(Queue {
+        queue: SegQueue::new(),
+        waker: AtomicWaker::new(),
+        liveness: AtomicUsize::new(SENDER_COUNT_BASE | RECEIVER_COUNT_BASE),
+        policy,
+    }));
+    (Sender { queue }, Receiver { queue })
+}
+
+/// A Collector Used in `BatchReceiver`.
+pub trait BatchCollector<Collection, Elem> {
+    /// If `elem` is collected into `collection` successfully, return `None`.
+    /// Otherwise return `elem` back, and `collection` should be spilled out.
+    fn collect(&mut self, collection: &mut Collection, elem: Elem) -> Option<Elem>;
+}
+
+pub struct VecCollector;
+
+impl<E> BatchCollector<Vec<E>, E> for VecCollector {
+    fn collect(&mut self, v: &mut Vec<E>, e: E) -> Option<E> {
+        v.push(e);
+        None
+    }
+}
+
+/// `BatchReceiver` is a `futures::Stream`, which returns a batched type.
+pub struct BatchReceiver<T, E, I, C> {
+    rx: Receiver<T>,
+    max_batch_size: usize,
+    elem: Option<E>,
+    initializer: I,
+    collector: C,
+}
+
+impl<T, E, I, C> BatchReceiver<T, E, I, C>
+where
+    T: Unpin,
+    E: Unpin,
+    I: Fn() -> E + Unpin,
+    C: BatchCollector<E, T> + Unpin,
+{
+    /// Creates a new `BatchReceiver` with given `initializer` and `collector`.
+    /// `initializer` is used to generate a initial value, and `collector`
+    /// will collect every (at most `max_batch_size`) raw items into the
+    /// batched value.
+    pub fn new(rx: Receiver<T>, max_batch_size: usize, initializer: I, collector: C) -> Self {
+        BatchReceiver {
+            rx,
+            max_batch_size,
+            elem: None,
+            initializer,
+            collector,
+        }
+    }
+}
+
+impl<T, E, I, C> Stream for BatchReceiver<T, E, I, C>
+where
+    T: Send + Unpin,
+    E: Unpin,
+    I: Fn() -> E + Unpin,
+    C: BatchCollector<E, T> + Unpin,
+{
+    type Item = E;
+
+    #[inline]
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let ctx = self.get_mut();
+        let (mut count, mut received) = (0, None);
+        let finished = loop {
+            // Our implementation is safe to poll twice.
+            match ctx.rx.poll_next_unpin(cx) {
+                Poll::Ready(Some(m)) => {
+                    let collection = ctx.elem.get_or_insert_with(&ctx.initializer);
+                    if let Some(m) = ctx.collector.collect(collection, m) {
+                        received = Some(m);
+                        break false;
+                    }
+                    count += 1;
+                    if count >= ctx.max_batch_size {
```

Comment:
If `max_batch_size` belongs to `BatchReceiver`, in which case we expect `BatchCollector::collect` returns `Some` and split?